File: dooropener/users_store.py
import json
import os
from datetime import datetime, timezone
from typing import Dict, Any, Optional


ISO_FORMAT = "%Y-%m-%dT%H:%M:%S%z"


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class UsersStore:
    """JSON-backed user store with cached reverse PIN lookup.

    - JSON schema:
      {
        "users": {
          "alice": {"pin": "1234", "active": true, "created_at": "...", "updated_at": "...", "last_used_at": null}
        }
      }
    - Maintains a cached ``{pin: username}`` reverse map, invalidated on any CRUD operation.
    """

    def __init__(self, path: str):
        self.path = path
        self.data: Dict[str, Any] = {"users": {}}
        self._loaded = False
        self._pin_cache: Dict[str, str] | None = None  # pin -> username

    def _load_file(self) -> None:
        if self._loaded:
            return
        try:
            if os.path.exists(self.path):
                with open(self.path, "r", encoding="utf-8") as f:
                    self.data = json.load(f)
                    if "users" not in self.data or not isinstance(
                        self.data["users"], dict
                    ):
                        self.data = {"users": {}}
            else:
                os.makedirs(os.path.dirname(self.path), exist_ok=True)
                self.data = {"users": {}}
        except Exception:
            self.data = {"users": {}}
        finally:
            self._loaded = True
            self._pin_cache = None  # invalidate on load

    def _save_atomic(self) -> None:
        os.makedirs(os.path.dirname(self.path), exist_ok=True)
        with open(self.path, "w", encoding="utf-8") as f:
            json.dump(self.data, f, ensure_ascii=False, indent=2)
        self._pin_cache = None  # invalidate on write

    def _invalidate_cache(self) -> None:
        self._pin_cache = None

    def get_pin_map(self) -> Dict[str, str]:
        """Return cached ``{pin: username}`` dict of active users."""
        self._ensure_loaded()
        if self._pin_cache is not None:
            return self._pin_cache
        result: Dict[str, str] = {}
        for user, meta in self.data.get("users", {}).items():
            if not bool(meta.get("active", True)):
                continue
            pin = meta.get("pin")
            if isinstance(pin, str) and 4 <= len(pin) <= 8 and pin.isdigit():
                result[pin] = user
        self._pin_cache = result
        return result

    def lookup_pin(self, pin: str) -> str | None:
        """Return username for *pin*, or ``None`` if no match."""
        return self.get_pin_map().get(pin)

    def list_users(self, include_pins: bool = False) -> Dict[str, Any]:
        self._ensure_loaded()
        items = []
        for user, meta in self.data.get("users", {}).items():
            item = {
                "username": user,
                "active": bool(meta.get("active", True)),
                "created_at": meta.get("created_at"),
                "updated_at": meta.get("updated_at"),
                "last_used_at": meta.get("last_used_at"),
                "times_used": meta.get("times_used", 0),
            }
            if include_pins:
                item["pin"] = meta.get("pin")
            items.append(item)
        return {"users": items}

    def _ensure_loaded(self) -> None:
        if not self._loaded:
            self._load_file()

    @staticmethod
    def _validate_username(username: str) -> bool:
        if not isinstance(username, str) or not (1 <= len(username) <= 32):
            return False
        allowed = set(
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-."
        )
        return all(c in allowed for c in username)

    @staticmethod
    def _validate_pin(pin: str) -> bool:
        return isinstance(pin, str) and pin.isdigit() and 4 <= len(pin) <= 8

    def create_user(self, username: str, pin: str, active: bool = True) -> None:
        self._ensure_loaded()
        if not self._validate_username(username):
            raise ValueError("Invalid username")
        if not self._validate_pin(pin):
            raise ValueError("Invalid pin")
        if username in self.data["users"]:
            raise KeyError("User already exists")
        now = _now_iso()
        self.data["users"][username] = {
            "pin": pin,
            "active": bool(active),
            "created_at": now,
            "updated_at": now,
            "last_used_at": None,
            "times_used": 0,
        }
        self._save_atomic()

    def update_user(
        self, username: str, pin: Optional[str] = None, active: Optional[bool] = None
    ) -> None:
        self._ensure_loaded()
        if username not in self.data["users"]:
            raise KeyError("User not found")
        if pin is not None and not self._validate_pin(pin):
            raise ValueError("Invalid pin")
        if active is not None:
            active = bool(active)
        meta = self.data["users"][username]
        if pin is not None:
            meta["pin"] = pin
        if active is not None:
            meta["active"] = active
        meta["updated_at"] = _now_iso()
        self._save_atomic()

    def delete_user(self, username: str) -> None:
        self._ensure_loaded()
        if username not in self.data["users"]:
            raise KeyError("User not found")
        del self.data["users"][username]
        self._save_atomic()

    def touch_user(self, username: str) -> None:
        self._ensure_loaded()
        if username in self.data["users"]:
            self.data["users"][username]["last_used_at"] = _now_iso()
            # Increment times_used counter, defaulting to 0 if not present (for existing users)
            self.data["users"][username]["times_used"] = (
                self.data["users"][username].get("times_used", 0) + 1
            )
            self._save_atomic()

    def user_exists(self, username: str) -> bool:
        self._ensure_loaded()
        return username in self.data["users"]

File: rust/src/users_store.rs
use anyhow::{bail, Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::io::Write;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserMeta {
    pub pin: String,
    pub active: bool,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub last_used_at: Option<String>,
    #[serde(default)]
    pub times_used: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
struct StoreData {
    users: HashMap<String, UserMeta>,
}

/// JSON-backed user store with atomic writes and merge-over-config behaviour.
///
/// This is a direct port of Python's `UsersStore`. Callers must hold the outer
/// `RwLock<UsersStore>` when accessing any method.
#[derive(Debug, Clone)]
pub struct UsersStore {
    path: PathBuf,
    data: StoreData,
}

impl UsersStore {
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self {
            path: path.into(),
            data: StoreData::default(),
        }
    }

    fn load(&mut self) {
        if self.path.exists() {
            match std::fs::read_to_string(&self.path) {
                Ok(contents) => {
                    self.data = serde_json::from_str(&contents).unwrap_or_default();
                }
                Err(_) => {
                    self.data = StoreData::default();
                }
            }
        } else {
            if let Some(parent) = self.path.parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            self.data = StoreData::default();
        }
    }

    fn save_atomic(&self) -> Result<()> {
        let dir = self.path.parent().unwrap_or(Path::new("."));
        std::fs::create_dir_all(dir)?;

        // Try to place the temp file on the same filesystem for atomic rename.
        let tmp_path = self.path.with_extension("tmp");
        let json = serde_json::to_string_pretty(&self.data)?;

        let mut file = std::fs::File::create(&tmp_path)
            .context("Failed to create temp file for users store")?;
        file.write_all(json.as_bytes())?;
        file.flush()?;
        drop(file);

        std::fs::rename(&tmp_path, &self.path).context("Failed to atomically replace users.json")?;
        Ok(())
    }

    /// Merge base_pins (from config.ini [pins]) with the JSON store.
    /// JSON store entries take precedence; inactive users are excluded.
    pub fn effective_pins(&mut self, base_pins: &HashMap<String, String>) -> HashMap<String, String> {
        self.load();
        let mut effective: HashMap<String, String> = base_pins.clone();
        for (username, meta) in &self.data.users {
            if !meta.active {
                effective.remove(username);
                continue;
            }
            if is_valid_pin(&meta.pin) {
                effective.insert(username.clone(), meta.pin.clone());
            }
        }
        effective
    }

    pub fn list_users(&mut self, include_pins: bool) -> Vec<serde_json::Value> {
        self.load();
        self.data
            .users
            .iter()
            .map(|(username, meta)| {
                let mut item = serde_json::json!({
                    "username": username,
                    "active": meta.active,
                    "created_at": meta.created_at,
                    "updated_at": meta.updated_at,
                    "last_used_at": meta.last_used_at,
                    "times_used": meta.times_used,
                });
                if include_pins {
                    item["pin"] = serde_json::Value::String(meta.pin.clone());
                }
                item
            })
            .collect()
    }

    pub fn create_user(&mut self, username: &str, pin: &str, active: bool) -> Result<()> {
        self.load();
        validate_username(username)?;
        validate_pin(pin)?;
        if self.data.users.contains_key(username) {
            bail!("User already exists");
        }
        let now = now_iso();
        self.data.users.insert(
            username.to_string(),
            UserMeta {
                pin: pin.to_string(),
                active,
                created_at: Some(now.clone()),
                updated_at: Some(now),
                last_used_at: None,
                times_used: 0,
            },
        );
        self.save_atomic()
    }

    pub fn update_user(&mut self, username: &str, pin: Option<&str>, active: Option<bool>) -> Result<()> {
        self.load();
        let meta = self
            .data
            .users
            .get_mut(username)
            .ok_or_else(|| anyhow::anyhow!("User not found"))?;

        if let Some(p) = pin {
            validate_pin(p)?;
            meta.pin = p.to_string();
        }
        if let Some(a) = active {
            meta.active = a;
        }
        meta.updated_at = Some(now_iso());
        self.save_atomic()
    }

    pub fn delete_user(&mut self, username: &str) -> Result<()> {
        self.load();
        if self.data.users.remove(username).is_none() {
            bail!("User not found");
        }
        self.save_atomic()
    }

    pub fn touch_user(&mut self, username: &str) {
        self.load();
        if let Some(meta) = self.data.users.get_mut(username) {
            meta.last_used_at = Some(now_iso());
            meta.times_used += 1;
            let _ = self.save_atomic();
        }
    }

    pub fn user_exists(&mut self, username: &str) -> bool {
        self.load();
        self.data.users.contains_key(username)
    }

    /// Return the username of an inactive user whose PIN matches, or None.
    pub fn find_disabled_user_by_pin(&mut self, pin: &str) -> Option<String> {
        self.load();
        for (username, meta) in &self.data.users {
            if !meta.active && constant_time_eq(pin, &meta.pin) {
                return Some(username.clone());
            }
        }
        None
    }
}

fn now_iso() -> String {
    chrono::Utc::now().to_rfc3339()
}

fn is_valid_pin(pin: &str) -> bool {
    pin.chars().all(|c| c.is_ascii_digit()) && pin.len() >= 4 && pin.len() <= 8
}

fn validate_pin(pin: &str) -> Result<()> {
    if !is_valid_pin(pin) {
        bail!("Invalid pin: must be 4-8 digits");
    }
    Ok(())
}

fn validate_username(username: &str) -> Result<()> {
    if username.is_empty() || username.len() > 32 {
        bail!("Invalid username: must be 1-32 characters");
    }
    if !username
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-' || c == '.')
    {
        bail!("Invalid username: only alphanumeric, _, -, . allowed");
    }
    Ok(())
}

/// Constant-time string equality to prevent timing attacks on PIN comparison.
fn constant_time_eq(a: &str, b: &str) -> bool {
    // Manual byte-by-byte compare that doesn't short-circuit (constant-time).
    if a.len() != b.len() {
        return false;
    }
    let mut diff: u8 = 0;
    for (x, y) in a.bytes().zip(b.bytes()) {
        diff |= x ^ y;
    }
    diff == 0
}
